tensornetwork/tests/network_components_free_test.py

```diff
@@ -316,6 +316,292 @@ def test_node_magic_matmul(backend):
+  assert node1.tensor.dtype == node2.tensor.dtype == result.dtype
+
+  if backend != 'tensorflow':
+    expected = np.array([[3, 4], [5, 6]])
+    result = (node1 + int_node).tensor
+    np.testing.assert_almost_equal(result, expected)
+    assert node1.tensor.dtype == int_node.tensor.dtype == result.dtype
+    expected = np.array([[3, 4], [5, 6]])
+    result = (int_node + node1).tensor
+    np.testing.assert_almost_equal(result, expected)
+    assert node1.tensor.dtype == int_node.tensor.dtype == result.dtype
+
+    expected = np.array([[3.5, 4.5], [5.5, 6.5]])
+    result = (node1 + float_node).tensor
+    np.testing.assert_almost_equal(result, expected)
+    assert float_node.dtype == result.dtype
+    expected = np.array([[3.5, 4.5], [5.5, 6.5]])
+    result = (float_node + node1).tensor
+    np.testing.assert_almost_equal(result, expected)
+    assert float_node.dtype == result.dtype
+
+
+def test_node_and_scalar_add_op(backend):
+  if backend != 'tensorflow':
+    node = Node(tensor=np.array([[1, 2], [3, 4]]), backend=backend)
+    expected = np.array([[3, 4], [5, 6]])
+    result = (node + 2).tensor
+    np.testing.assert_almost_equal(result, expected)
+    assert node.tensor.dtype == result.dtype
+
+    expected = np.array([[3.5, 4.5], [5.5, 6.5]])
+    np.testing.assert_almost_equal((node + 2.5).tensor, expected)
+
+    node = Node(tensor=np.array([[1., 2.], [3., 4.]]), backend=backend)
+    result = (node + 2.5).tensor
+    np.testing.assert_almost_equal(result, expected)
+    assert node.tensor.dtype == result.dtype
+
+
+def test_between_node_sub_op(backend):
+  node1 = Node(tensor=np.array([[1, 2], [3, 4]]), backend=backend)
+  node2 = Node(tensor=np.array([[10, 10], [10, 10]]), backend=backend)
+  int_node = Node(tensor=2, backend=backend)
+  float_node = Node(tensor=2.5, backend=backend)
+
+  expected = np.array([[-9, -8], [-7, -6]])
+  result = (node1 - node2).tensor
+  np.testing.assert_almost_equal(result, expected)
+  assert node1.tensor.dtype == node2.tensor.dtype == result.dtype
+
+  if backend != 'tensorflow':
+    expected = np.array([[-1, 0], [1, 2]])
+    result = (node1 - int_node).tensor
+    np.testing.assert_almost_equal(result, expected)
+    assert node1.tensor.dtype == int_node.tensor.dtype == result.dtype
+    expected = np.array([[1, 0], [-1, -2]])
+    result = (int_node - node1).tensor
+    np.testing.assert_almost_equal(result, expected)
+    assert node1.tensor.dtype == int_node.tensor.dtype == result.dtype
+
+    expected = np.array([[-1.5, -0.5], [0.5, 1.5]])
+    result = (node1 - float_node).tensor
+    np.testing.assert_almost_equal(result, expected)
+    assert float_node.dtype == result.dtype
+    expected = np.array([[1.5, 0.5], [-0.5, -1.5]])
+    result = (float_node - node1).tensor
+    np.testing.assert_almost_equal(result, expected)
+    assert float_node.dtype == result.dtype
+
+
+def test_node_and_scalar_sub_op(backend):
+  if backend != 'tensorflow':
```

Comment:
Interesting, Why are you always excluding the TF backend? Are you testing with TF1 or TF2?

Reply:
I'm testing on **TF2.1.0**


I posted the reason why I did it at the upper comment. TF doesn't accept algebraic between different dtype like `int32` with `int64` or `int64` with `float64` different from numpy or PyTorch. **So if I put a scalar value on the Tensor the TF will throw an error.**

    tensor(2) <- int32
    tensor(2.) <- float32
    tensor([2,2]) <- int64
    tensor([2.,2.]) <- float64

So I had to exclude the test that is held on different dtype in the case of TF. If you see in the test cases I allow TF to test when the dtype is the same.


I thought about making a test case especially for TF that raise an error with different dtype calculation but I think rather than that it would be good to support BaseNode to accept the dtype as a attribute :) So we can specify the TF tensor dtype so can handle the calculation with scalar type.

Or, can make a special `implicit dtype conversion for TF tensor` to **TF backend** that check the dtypes or operands and automatically make a conversion to make different dtype same.

    int32, int64 -> int64, int64
    int32, float64 -> float64, float64

In this case, I could get off that test case that TF excluding part.


If I am misunderstanding something please tell me :)

Reply:
Hey @IllgamhoDuck, first of all, thanks for the effort.
If a test needs to be skipped, can you use `pytest.skip`with an appropriate reason, it makes it easier for later follow up and overall housekeeping?

Reply:
Even easier would be to just wrap all of your scalars in a `np.asarray`. I prefer that over skipping tests.

Reply:
@MichaelMarien Thks for the advice! I will next time :) I was missing the important point! 

@Thenerdstation Got it 👍 I will change the `dtype` with `np.asarray` for TF but will leave other backends just as it is right now.